src/Model/Variant/Collection.php

```diff
@@ -0,0 +1,189 @@
+
+    /**
+     * @var array
+     */
+    private $childrenMap = [];
+
+    /**
+     * @var string[]
+     */
+    private $attributeCodes = [];
+
+    /**
+     * @param CollectionFactory $childCollectionFactory
+     * @param ProductFactory $productFactory
+     * @param SearchCriteriaBuilder $searchCriteriaBuilder
+     * @param DataProvider $productDataProvider
+     * @param MetadataPool $metadataPool
+     */
+    public function __construct(
+        CollectionFactory $childCollectionFactory,
+        ProductFactory $productFactory,
+        SearchCriteriaBuilder $searchCriteriaBuilder,
+        DataProvider $productDataProvider,
+        MetadataPool $metadataPool
+    ) {
+        $this->childCollectionFactory = $childCollectionFactory;
+        $this->productFactory = $productFactory;
+        $this->searchCriteriaBuilder = $searchCriteriaBuilder;
+        $this->productDataProvider = $productDataProvider;
+        $this->metadataPool = $metadataPool;
+    }
+
+    /**
+     * Add parent to collection filter
+     *
+     * @param Product $product
+     * @return void
+     */
+    public function addParentProduct(Product $product) : void
+    {
+        $linkField = $this->metadataPool->getMetadata(ProductInterface::class)->getLinkField();
+        $productId = $product->getData($linkField);
+
+        if (isset($this->parentProducts[$productId])) {
+            return;
+        }
+
+        if (!empty($this->childrenMap)) {
+            $this->childrenMap = [];
+        }
+        $this->parentProducts[$productId] = $product;
+    }
+
+    /**
+     * Add attributes to collection filter
+     *
+     * @param array $attributeCodes
+     * @return void
+     */
+    public function addEavAttributes(array $attributeCodes) : void
+    {
+        $this->attributeCodes = array_replace($this->attributeCodes, $attributeCodes);
+    }
+
+    /**
+     * Retrieve child products from for passed in parent id.
+     *
+     * @param int $id
+     * @return array
+     */
+    public function getChildProductsByParentId(int $id) : array
+    {
+        $childrenMap = $this->fetch();
+
+        if (!isset($childrenMap[$id])) {
+            return [];
+        }
+
+        return $childrenMap[$id];
+    }
+
+    /**
+     * Fetch all children products from parent id's.
+     *
+     * @return array
+     * @throws \Exception
+     */
+    private function fetch() : array
+    {
+        if (empty($this->parentProducts) || !empty($this->childrenMap)) {
+            return $this->childrenMap;
+        }
+
+        foreach ($this->parentProducts as $product) {
+            $attributeData = $this->getAttributesCodes($product);
+            /** @var ChildCollection $childCollection */
+            $childCollection = $this->childCollectionFactory->create();
+            $childCollection->setProductFilter($product);
+            $childCollection->addAttributeToSelect($attributeData);
+            $childCollection->addAttributeToFilter('status', '1');
```

Comment:
Please use constants whenever possible, for referencing "status enabled" you may use:
vendor/magento/module-catalog/Model/Product/Attribute/Source/Status.php ::STATUS_ENABLED

Reply:
Done